### PR TITLE
Fix typo in IMapFunction::backtrace

### PR DESCRIPTION
**Pull request by niltonlk**
Executing some mpi tests in debug mode (e.g. sli -d mpitests/test_all_to_all.sli) resulted in

sli: nest-simulator/sli/sliarray.cc:944: void SLIArrayModule::IMapFunction::backtrace(SLIInterpreter *, int) const: Assertion `count == NULL' failed.

Since line 949 access variable count, the assertion should not be NULL.
Changing == to != in line 944, the error disappeared.

This is also somewhat related to issue #784, where the same message was observed.

### CODE REVIEW

**Review by jougs — APPROVED** [review]
This seems reasonable to me. Thanks for addressing this. As this PR fixes the problem and also only pertains to simulations running in debug mode, I'm merging without a second review.

**Review by jougs — CHANGES_REQUESTED** [review]
Argh! I should have looked at the Travis log more closely before approving. It seems that the change actually makes an MPI test fail rather than pass. @niltonlk: can you please have another look and check what's going on?

**Comment by niltonlk** [comment]
This was the installcheck error in Travis log:

> Running '/home/travis/build/nest/nest-simulator/result/bin/sli' '/home/travis/build/nest/nest-simulator/result/share/doc/nest/mpitests/test_multiple_synapses_spatial_networks.sli'
SLI HEAD@3d74f3f4 (C) 2004 The NEST Initiative
N_MPI: 1
N_MPI: 2
[travis-job-86d97fd0-a495-4e8e-bf38-31b9faf35dba:15376] 1 more process has sent help message help-mpi-btl-base.txt / btl:no-nics
[travis-job-86d97fd0-a495-4e8e-bf38-31b9faf35dba:15376] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
N_MPI: 4
Assertion false failed.
Nov 02 07:39:38 assert [Error]: FailedAssertion

Is it OK to just undo the change to see if the travis check passes? 

I have also just checked that there are other backtrace functions with assert( count == NULL ) and others with ( count != NULL ). They are:

**assert( count == NULL )**
- line 938: SLIArrayModule::IMapFunction::backtrace( SLIInterpreter* i, int p ) const (this is the block I have "corrected")
- line 1058: SLIArrayModule::IMap_ivFunction::backtrace( SLIInterpreter* i, int p ) const
- line 1197: SLIArrayModule::IMap_dvFunction::backtrace( SLIInterpreter* i, int p ) const

**assert( count != NULL )**
- line 1436: SLIArrayModule::IMapIndexedFunction::backtrace( SLIInterpreter* i, int p ) const
- line 1588: SLIArrayModule::IMapThreadFunction::backtrace( SLIInterpreter* i, int p ) const
- line 4006: SLIArrayModule::Iforall_ivFunction::backtrace( SLIInterpreter* i, int p ) const
- line 4082: SLIArrayModule::Iforall_dvFunction::backtrace( SLIInterpreter* i, int p ) const

I believe there are two more places that need correction, as all of those blocks access count on the same way. I believe the three blocks which have count == NULL are all typos. Anyway, I am digging on this issue and will get back when got more outputs.

**Comment by heplesser** [comment]
@niltonlk Could you on your own computer run the following in the `testsuite/mpitests` directory:
```
mpirun -np 4 nest test_multiple_synapses_spatial_networks.sli
```
and also try it with 1 and 2 instead of 4?  In that way we could get more information on precisely which assertion failed. You could also try with the `-d` flag to activate debugging?

**Comment by niltonlk** [comment]
@heplesser Here follows the 6 outputs of:

`mpirun -np 4 nest test_multiple_synapses_spatial_networks.sli`

with np = 4, 2, 1 and debug ON/OFF:

**debug OFF**
1. `mpirun -np 4 nest test_multiple_synapses_spatial_networks.sli`
[test_multiple_synapses_spatial_networks_4.txt](https://github.com/nest/nest-simulator/files/5488016/test_multiple_synapses_spatial_networks_4.txt)

2. `mpirun -np 2 nest test_multiple_synapses_spatial_networks.sli`
[test_multiple_synapses_spatial_networks_2.txt](https://github.com/nest/nest-simulator/files/5488022/test_multiple_synapses_spatial_networks_2.txt)

3. `mpirun -np 1 nest test_multiple_synapses_spatial_networks.sli`
[test_multiple_synapses_spatial_networks_1.txt](https://github.com/nest/nest-simulator/files/5488025/test_multiple_synapses_spatial_networks_1.txt)

**debug ON**
4. `mpirun -np 4 nest -d test_multiple_synapses_spatial_networks.sli`
[test_multiple_synapses_spatial_networks_4_debug.txt](https://github.com/nest/nest-simulator/files/5488033/test_multiple_synapses_spatial_networks_4_debug.txt)

5. `mpirun -np 2 nest -d test_multiple_synapses_spatial_networks.sli`
[test_multiple_synapses_spatial_networks_2_debug.txt](https://github.com/nest/nest-simulator/files/5488036/test_multiple_synapses_spatial_networks_2_debug.txt)

6. `mpirun -np 1 nest -d test_multiple_synapses_spatial_networks.sli`
[test_multiple_synapses_spatial_networks_1_debug.txt](https://github.com/nest/nest-simulator/files/5488041/test_multiple_synapses_spatial_networks_1_debug.txt)

I have set OMPI_MCA_btl=varder,self,tcp, which is the same as invoking mpirun with --mca btl vader,self,tcp
vader is the replacement of sm, which stands for shared memory
the first few lines with "unable to open mca_mtl_ofi" can be omitted by invoking mpirun with --mca mtl ^ofi

In all condition, the test was successfull in my local configuration.
I believe it is more a travis configuration issue, as the message suggests that it could not find a NIC for communication.

I will try to configure travis on my local repository so that I can look for a solution.

**Comment by niltonlk** [comment]
These corrections resulted in no Linux-GCC errors in my local travis environment.

**Comment by heplesser** [comment]
@niltonlk If the failed MPI test on Travis earlier was due to a missing NIC, that may have been a temporary problem on Travis, since all other MPI tests passed in that run, i.e., found a NIC. Also, in the most recent version all tests pass as expected. The failure on macOS is due to changes in Python setup there, see #1822. 

We recently discovered one subtle error (lack of intialization of a variable) which surfaced only under macOS, so I'd rather wait for #1822 to be fixed before approving this PR, so we have all tests passed before we merge.

**Review by heplesser — APPROVED** [review]
Thanks. Approving the changes as such, merging conditional on all tests passing once #1822 is fixed.

**Comment by jougs** [comment]
Did I miss something? Why was this closed and the branch deleted _before_ it was merged?

**Comment by niltonlk** [comment]
Sorry, I was deleting some branches I used to test my local travis configuration, i ended up deleting this branch by mistake. Will restore it...

**Comment by jougs** [comment]
OK. Thanks!

**Review by jougs — APPROVED** [review]
Thanks for taking care. Approving. And this time I mean it ;-)

**Pull request merged**
